### ISSUE TITLE
Custom polynomials not working because crc-rs does not implement the complete CRC32 algorithm.

### ISSUE THREAD

**Issue by CLomanno** [mrhooray/crc-rs#31]
Is there a reason you chose to implement the CRC32 algorithm in a reversed order?

Everything from defining the polynomial to generating the lookup table to updating the CRC is reversed.

In `build.rs` IEEE is defined as 0xedb88320 which is the reflected form of the correct polynomial 0x04C11DB7.  Then `make_table_crc32()` works in a Little Endian way.

I am currently using a custom polynomial and custom XOR value, and I cannot use your code in the current state.  I have modified it to use the correct CRC32 algorithm.  Are you interested in me making a pull request?

**Comment by mrhooray** [comment]
Hey @CLomanno I started this crate as an exercise by following Golang's implementation of CRC[32](https://golang.org/src/hash/crc32/crc32.go)/[64](https://golang.org/src/hash/crc64/crc64.go). I'm by no means an expert on this topic and assumed they have good reasons to do so.

Would picking a proper polynomial with matching initial through `new_with_initial` work for your scenarios?

**Comment by CLomanno** [comment]
`new_with_initial()` would not work because my CRC is not reflected.  That's why I added the parameter `ref` to the `make_table()` and `update()` functions.

The way you currently calculate all the CRC values with a `!value` at the beginning and end is a shortcut that *only* if your initial value is 0xFFFFFFFF and your final XOR value is 0xFFFFFFFF.

Please see my pull request (#32) for code that works for all 32bit CRC cases.  It is fully backwards compatable with your current version of 1.7.0.  I also put 2 good links in there that describe the various 32bit CRC standards, polynomials, initial values, and XORs.

**Comment by CLomanno** [comment]
Your right, both *normal* and *reversed* can lead to the same result, but since you use the table lookup method and you force the user to use the *reversed* lookup table, the user can only find a *reversed* solution.  

The math is different for forward and backward crc-32 algorithms.  Look what I added:
https://github.com/CLomanno/crc-rs/blob/f44dc5d7e984c8dca6eee333edb346d8f8265235/src/crc32.rs#L36-L44

Your code is up top in the original implementation inside `if rfl==true`.  My unreflected math is in the else statement.

Also, I kept your simple implimentation of `new()` and `new_with_initial()`.  I just added support for custom polynomials.

**Comment by mrhooray** [comment]
@CLomanno thanks for digging into this and updating #32 

a few questions since you seem to understand this topic quite well:
- if both normal and reversed approach with proper math leads to correct result with same flexibility, shall we only choose one to keep things simple?
- thoughts on keep implementation between 16/32/64 same/consistent(in terms of normal or reversed , plus final xor)?

i'm onboard with ideas that leads to a simple, consistent and **correct/complete**(what this thread is about) implementation, even if breaking changes are involved.

**Comment by CLomanno** [comment]
@mrhooray 
You have 250k downloads of this crate, people use it.

To make it as usable as possible I think you should support both normal and reversed.  You also need to support a final xor value.

`pub struct Digest {`
`    table: [u32; 256],`
`    initial: u32,`
`    value: u32,`
`    reflect: bool,`
`    final_xor: u32,`
`}`
I updated the struct `Digest` to handle everything.  My code is 100% reverse comparable with yours.  There are still the simple `new()` and `new_with_initial()` and the front end hasn't changed.

I agree with you, though.  CRC16 and CRC32 should be updated to have the same structure.

As of right now, the CRC32 code is complete and correct.

**Comment by mrhooray** [comment]
> You have 250k downloads of this crate, people use it.

not saying we'd break people automatically. major version bump should help avoid that

> CRC16 and CRC64 should be updated to have the same structure.

i'll play around with #32 and we can merge it to unblock people needing the added capability

would you also be able to help on updating crc16/64 as next steps? seems your expertise could help restructure internal implementation and shape roadmap towards 2.0(if we're thinking about using non-reversed approach only)

**Comment by nickbabcock** [comment]
Hey I believe this issue is somewhat relevant to me. Here's a comment in [my own crc implementation](https://github.com/nickbabcock/boxcars/blob/e2007bdb1fde9cfe4b4e94e1638ec2adf84ca7ab/src/crc.rs) (unwrapped quote and code for readability):

---

Calculates the crc-32 for rocket league replays. Not all CRC algorithms are the same. The crc
algorithm can be generated with the following parameters (pycrc):

- `Width` = 32
- `Poly` = 0x04c11db7
- `XorIn` = 0x10340dfe
- `ReflectIn` = False
- `XorOut` = 0xffffffff
- `ReflectOut` = False

It may be tempting to optimize this function to use slice by techniques, but I have not be able
to achieve this. pycrc disables slice by for non-reflected algorithms, and I couldn't get the
reversed model from reveng to pass the tests. I've wasted too much time trying to optimize this
function. I thought it would be worthwhile as when one is only interested in parsing a replay's
header and checking the crc, the crc computation is 100x more intensive than the header
parsing.

I've even tried to lookup the slice by 8 tables of `CRCTablesSB8` and
`CRCTablesSB8_DEPRECATED` from the unreal engine to glean any information on derivation or
usage. They can be found in the `AHRUnrealEngine` Github repo. I've copied them and the usage
in `MemCrc_DEPRECATED` faithfully, but no luck. This has been a teachable moment

**Comment by CLomanno** [comment]
@mrhooray I got the crc64 working like crc32.  I'll get crc16 updated soon.

@nickbabcock The code I have been working on takes in all of the parameters you mentioned except width.  Right now there is a lot of duplicated code for crc16, crc32, and crc64.  Width would be a nice way to clean up the implementations.  Currently my code calculates both normal and reflected CRCs.  It's close to being completed.

**Comment by mrhooray** [comment]
we've landed #32 and #35 however `new_with_initial` seems to contain breaking change as commented in #35 

gonna postpone release still this has been discussed

**Comment by CLomanno** [comment]
@mrhooray 
I branched and made an update in #36

**Comment by mrhooray** [comment]
merged #36 #38
